x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,431 @@
+				close(c)
+				return nil
+			case err := <-errCh:
+				// Handle errors received from goroutines
+				m.Logger().Error(err.Error())
+			case healthDetails, ok := <-c:
+				if !ok {
+					return nil
+				}
+				m.Logger().Debugf("[AWS Health] [DescribeEventDetails] Event ARN : %s, Affected Entities (Pending) : %d, Affected Entities (Resolved): %d, Affected Entities (Others) : %d", *healthDetails.event.Arn, healthDetails.affectedEntityPending, healthDetails.affectedEntityResolved, healthDetails.affectedEntityOthers)
+				events = append(events, createEvents(healthDetails))
+			}
+		}
+		if eventOutput.NextToken == nil {
+			break
+		}
+		nextTokenString = *eventOutput.NextToken
+	}
+	close(c)
+	close(errCh)
+	return events
+}
+
+// createEvents takes in a HealthDetails struct and returns an mb.Event struct
+// populated with the data from the HealthDetails. It sets the MetricSetFields,
+// RootFields and ID fields of the mb.Event struct. The MetricSetFields contain
+// the details of the health event. The RootFields specify that it is an AWS
+// event. The ID is generated from the event ARN, status code and current date.
+func createEvents(hd HealthDetails) mb.Event {
+	currentDate := getCurrentDateTime()
+	eventID := currentDate + getValueOrDefault(hd.event.Arn, "") + getValueOrDefault((*string)(&hd.event.StatusCode), "")
+	event := mb.Event{
+		MetricSetFields: mapstr.M{
+			"event_arn":                  getValueOrDefault(hd.event.Arn, ""),
+			"end_time":                   getValueOrDefault(hd.event.EndTime, time.Time{}),
+			"event_scope_code":           getValueOrDefault((*string)(&hd.event.EventScopeCode), ""),
+			"event_type_category":        getValueOrDefault((*string)(&hd.event.EventTypeCategory), ""),
+			"event_type_code":            getValueOrDefault(hd.event.EventTypeCode, ""),
+			"last_updated_time":          getValueOrDefault(hd.event.LastUpdatedTime, time.Time{}),
+			"region":                     getValueOrDefault(hd.event.Region, ""),
+			"service":                    getValueOrDefault(hd.event.Service, ""),
+			"start_time":                 getValueOrDefault(hd.event.StartTime, time.Time{}),
+			"status_code":                getValueOrDefault((*string)(&hd.event.StatusCode), ""),
+			"affected_entities_pending":  hd.affectedEntityPending,
+			"affected_entities_resolved": hd.affectedEntityResolved,
+			"affected_entities_others":   hd.affectedEntityOthers,
+			"affected_entities":          createAffectedEntityDetails(hd.affectedEntities),
+			"event_description":          hd.eventDescription,
+		},
+		RootFields: mapstr.M{
+			"cloud.provider": "aws",
+		},
+		ID: generateEventID(eventID),
+	}
+	return event
+}
+
+type HealthDetails struct {
+	event                  types.Event
+	eventDescription       string
+	affectedEntities       []types.AffectedEntity
+	affectedEntityPending  int32
+	affectedEntityResolved int32
+	affectedEntityOthers   int32
+}
+
+type AffectedEntityDetails struct {
```

Comment:
is it more efficient to just have this as a `mapstr.M` which is returned from `createAffectedEntityDetails`?